src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.cpp

```diff
@@ -0,0 +1,387 @@
+                                                      int32_t* pOSStatus)
+{
+    if (pOSStatus != nullptr)
+        *pOSStatus = noErr;
+
+    if (sslContext == nullptr || pszTargetName == nullptr || pOSStatus == nullptr)
+        return -1;
+
+    if (cbTargetName < 0)
+        return -2;
+
+    size_t currentLength;
+    *pOSStatus = SSLGetPeerDomainNameLength(sslContext, &currentLength);
+
+    // We'll end up walking down the path that sets the hostname more than once during
+    // the handshake dance. But once the handshake starts Secure Transport isn't willing to
+    // listen to this.  So, if we've already set it, don't set it again.
+    if (*pOSStatus == noErr && currentLength == 0)
+    {
+        *pOSStatus = SSLSetPeerDomainName(sslContext, pszTargetName, static_cast<size_t>(cbTargetName));
+    }
+
+    return *pOSStatus == noErr;
+}
+
+extern "C" int32_t
+AppleCryptoNative_SslSetIoCallbacks(SSLContextRef sslContext, SSLReadFunc readFunc, SSLWriteFunc writeFunc)
+{
+    return SSLSetIOFuncs(sslContext, readFunc, writeFunc);
+}
+
+extern "C" PAL_TlsHandshakeState AppleCryptoNative_SslHandshake(SSLContextRef sslContext)
+{
+    if (sslContext == nullptr)
+        return PAL_TlsHandshakeState_Unknown;
+
+    OSStatus osStatus = SSLHandshake(sslContext);
+
+    switch (osStatus)
+    {
+        case noErr:
+            return PAL_TlsHandshakeState_Complete;
+        case errSSLWouldBlock:
+            return PAL_TlsHandshakeState_WouldBlock;
+        case errSSLServerAuthCompleted:
+            return PAL_TlsHandshakeState_ServerAuthCompleted;
+        default:
+            return osStatus;
+    }
+}
+
+static PAL_TlsIo OSStatusToPAL_TlsIo(OSStatus status)
+{
+    switch (status)
+    {
+        case noErr:
+            return PAL_TlsIo_Success;
+        case errSSLWouldBlock:
+            return PAL_TlsIo_WouldBlock;
+        case errSSLClosedGraceful:
+            return PAL_TlsIo_ClosedGracefully;
+        default:
+            return status;
+    }
+}
+
+extern "C" PAL_TlsIo
+AppleCryptoNative_SslWrite(SSLContextRef sslContext, const uint8_t* buf, uint32_t bufLen, uint32_t* bytesWritten)
+{
+    if (bytesWritten == nullptr)
+        return PAL_TlsIo_Unknown;
+
+    size_t expected = static_cast<size_t>(bufLen);
+    size_t totalWritten;
+
+    OSStatus status = SSLWrite(sslContext, buf, expected, &totalWritten);
+
+    if (status != noErr)
+    {
+        *bytesWritten = static_cast<uint32_t>(totalWritten);
+        return OSStatusToPAL_TlsIo(status);
+    }
+
+    return PAL_TlsIo_Success;
+}
+
+extern "C" PAL_TlsIo
+AppleCryptoNative_SslRead(SSLContextRef sslContext, uint8_t* buf, uint32_t bufLen, uint32_t* written)
+{
+    if (written == nullptr)
+        return PAL_TlsIo_Unknown;
+
+    size_t writtenSize = 0;
+    size_t bufSize = static_cast<size_t>(bufLen);
+
+    OSStatus status = SSLRead(sslContext, buf, bufSize, &writtenSize);
+    *written = static_cast<uint32_t>(writtenSize);
```

Comment:
Currently we only support amd64 for macOS.  I don't know if we'd ever be supporting x86 (or arm32, if they have one), but I think taking the conservative (32-bit) approach is safer.

I'm hardening this a bit to ensure that we don't lose data... though if `SSLRead` doesn't respect bufSize (a uint32_t) as a boundary then there's not a lot we can do for application safety.